src/execution/service.ts

```diff
@@ -109,22 +137,148 @@ export class ServiceScriptExecution extends BaseExecutionWithCommand<ServiceScri
    * execute the command at this stage in the case of services.
    */
   protected override async _execute(): Promise<ExecutionResult> {
-    const dependencyFingerprints = await this._executeDependencies();
-    if (!dependencyFingerprints.ok) {
-      return dependencyFingerprints;
+    switch (this._state.id) {
+      case 'initial': {
+        this._state = {id: 'fingerprinting'};
+        const dependencyFingerprints = await this._executeDependencies();
+        if (!dependencyFingerprints.ok) {
+          return dependencyFingerprints;
+        }
+        const fingerprint = await Fingerprint.compute(
+          this._config,
+          dependencyFingerprints.value
+        );
+        this._state = {id: 'unstarted'};
```

Comment:
I think it's a little confusing to have the state set twice in this case. It's also a bit confusing to see some of these functions be async and use await and others use unawaited promises to set up the next transition... along with the boilerplate of the non-relevant states, I'd maybe in the future consider a state machine library, or some helpers to make the machine even more declarative.

Reply:
True. Replaced the `async` function with the event method style. This does require another state, but it makes sense. (The one other `async` function was just a mistake, it didn't actually `await`).

On the boilerplate: as this class grows, a lot more cases end up being meaningful. For now, I like that it forces me to consider every possible state. In some cases there will end up being only one valid state, so that could just be replaced with a simple assert at the top. Keeping it all consistent for now though.